src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs

```diff
@@ -119,60 +119,25 @@ public ZipArchive(Stream stream, ZipArchiveMode mode, bool leaveOpen) : this(str
         /// </param>
         /// <exception cref="ArgumentException">If a Unicode encoding other than UTF-8 is specified for the <code>entryNameEncoding</code>.</exception>
         public ZipArchive(Stream stream, ZipArchiveMode mode, bool leaveOpen, Encoding? entryNameEncoding)
+            : this(mode, leaveOpen, entryNameEncoding, backingStream: null, archiveStream: null)
         {
             ArgumentNullException.ThrowIfNull(stream);
 
-            EntryNameAndCommentEncoding = entryNameEncoding;
             Stream? extraTempStream = null;
 
             try
             {
                 _backingStream = null;
 
-                // check stream against mode
-                switch (mode)
+                if (ValidateMode(mode, stream))
                 {
-                    case ZipArchiveMode.Create:
-                        if (!stream.CanWrite)
-                            throw new ArgumentException(SR.CreateModeCapabilities);
-                        break;
-                    case ZipArchiveMode.Read:
-                        if (!stream.CanRead)
-                            throw new ArgumentException(SR.ReadModeCapabilities);
-                        if (!stream.CanSeek)
-                        {
-                            _backingStream = stream;
-                            extraTempStream = stream = new MemoryStream();
-                            _backingStream.CopyTo(stream);
-                            stream.Seek(0, SeekOrigin.Begin);
-                        }
-                        break;
-                    case ZipArchiveMode.Update:
-                        if (!stream.CanRead || !stream.CanWrite || !stream.CanSeek)
-                            throw new ArgumentException(SR.UpdateModeCapabilities);
-                        break;
-                    default:
-                        // still have to throw this, because stream constructor doesn't do mode argument checks
-                        throw new ArgumentOutOfRangeException(nameof(mode));
+                    _backingStream = stream;
+                    extraTempStream = stream = new MemoryStream();
+                    _backingStream.CopyTo(stream);
+                    stream.Seek(0, SeekOrigin.Begin);
```

Comment:
I have the blurry idea that this can be in a helper `DecideBackingStream` and also moved to the `this :` call next to archiveStream.

Reply:
Discussed via chat, might be too complex.